### PR TITLE
 HPCC-26921 Avoid CQ's re-looking up logical files.

### PR DESCRIPTION
**Pull request by jakesmith**
If a CQ performed a logical file lookup, e.g. via a diskread/kj/
indexread etc., and the logical filename could vary, e.g. it was
computed each time, then each worker would re-initialized the
manager activity, and cause a new Dali lookup per iteration.
e.g. on a 400 way Thor executing a CQ 1000 times, there would be
400,000 Dali lookups.

Change so that if within a CQ, the DFS lookups are cached
(not cleared) per execution. This results in the LFN being looked
up only once if the computed logical file remains constant.

Also, add temporary workaround for subFileStats tracking issues,
related to activities being able to have >1 readfile, some of
which can be superfiles.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review/discuss. This I believe is the main reason for severe slowness in the reported JIRA (and others similar).
There are some incidental changes to the subfile stats stuff, but spotted a few issues in relation to queries like these, but also KJ's (where>1 file read) that need addressing, but in a separate JIRA.

File: thorlcr/graph/thgraphmaster.cpp [hunk 1]
```diff
@@ -542,35 +556,84 @@ void CMasterActivity::getEdgeStats(IStatisticGatherer & stats, unsigned idx)
...
 void CMasterActivity::updateFileReadCostStats(std::vector<OwnedPtr<CThorStatsCollection>> & subFileStats)
 {
-    if (!subFileStats.empty())
-    {
-        unsigned numSubFiles = subFileStats.size();
-        for (unsigned i=0; i<numSubFiles; i++)
-        {
-            IDistributedFile *file = queryReadFile(i);
-            if (file)
+    /* JCSMORE->SHAMSER: (separate JIRA needed)
```
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp:584]: Opened JIRA: https://track.hpccsystems.com/browse/HPCC-27399

**Review by ghalliday — APPROVED** [review]
@jakesmith a few trivial comments.  will defer to @shamser about the stats related changes - although they looked good to me.

**Comment by ghalliday** [comment]
@shamser please note the comments jake has added for you in the PR, and can you double check his related changes

**Review by shamser — COMMENTED** [review]
I see that subFileStats will be expanded at least to the size of the numsubs.  However, if the numsubs is smaller in a subsequent CQ execution,  it wouldn't get shrunk back.  If that is the case, wouldn't the workers deserialize too few CThorStatsCollection as the manager's array would be larger?

**Comment by jakesmith** [comment]
> I see that subFileStats will be expanded at least to the size of the numsubs. However, if the numsubs is smaller in a subsequent CQ execution, it wouldn't get shrunk back. If that is the case, wouldn't the workers deserialize too few CThorStatsCollection as the manager's array would be larger?

yes, subFileStats will the high water mark, but any worker might be processing a file with less subfiles, meaning it will send less, and there'll be deserialization error as it tries to deserialize more than is available.
The full fix is going to require a map, or table in the manager, and the workers will need to identify which file (subfile) specifically they are dealing with/sending back , so the manager can reference the correct CThorStatsCollection.

I don't think it's worth going any further in this PR, it can still hit deserialization errors if there are variable filenames that have mix of super sizes (or some might not be supers), but it's why I opened up the separate JIRA: https://track.hpccsystems.com/browse/HPCC-27399

**Review by shamser — APPROVED** [review]

**Pull request merged**